TrkDiag/src/TrackResolution_module.cc

```diff
@@ -0,0 +1,166 @@
+    bool _makeHists;
+    int _debug;
+
+    enum {kFirst, kSecond, kNeither}; // possible resolution results
+
+    struct Hist_t {
+      TH1* fitcon;
+    };
+    Hist_t _hists;
+  };
+
+  //--------------------------------------------------------------------------------------------------------
+  TrackResolution::TrackResolution(const Parameters& conf) :
+    art::EDProducer{conf},
+    _seedTag(conf().seedTag()),
+    _useCaloHit(conf().useCaloHit()),
+    _makeHists(conf().makeHists()),
+    _debug(conf().debug())
+  {
+    produces<KalSeedPtrCollection>();
+    if(_makeHists) {
+      art::ServiceHandle<art::TFileService> tfs;
+      art::TFileDirectory tfdir = tfs->mkdir("resolving");
+      _hists.fitcon = tfdir.make<TH1F>("fitcon", "Track p(#chi^2)", 200,  0.,  1.);
+    }
+  }
+
+  //--------------------------------------------------------------------------------------------------------
+  // Decide which track is better
+  int TrackResolution::resolve(const KalSeed* k_1, const KalSeed* k_2) {
+    if(!k_1 && !k_2) return kNeither;
+    if(!k_1) return kSecond;
+    if(!k_2) return kFirst;
+
+    const float fitcon_1(k_1->fitConsistency()), fitcon_2(k_2->fitConsistency());
+    if(_debug > 1) printf("    Comparing two tracks: Calo cluster: k1 = %o, k2 = %o; Fit quality: k1 = %.3g, k2 = %.3g\n",
+                          k_1->hasCaloCluster(), k_2->hasCaloCluster(), fitcon_1, fitcon_2);
+
+    // If requested, prioritize tracks that include a calorimeter cluster
+    if(_useCaloHit ) {
+      if( k_1->hasCaloCluster() && !k_2->hasCaloCluster()) return kFirst;
+      if(!k_1->hasCaloCluster() &&  k_2->hasCaloCluster()) return kSecond;
+    }
+
+    // Compare the fit quality
+    if(fitcon_1 > fitcon_2) return kFirst;
+    return kSecond;
+  }
+
+  //--------------------------------------------------------------------------------------------------------
+  void TrackResolution::produce(art::Event& event ) {
+    // create output
+    std::unique_ptr<KalSeedPtrCollection> trkcol(new KalSeedPtrCollection());
+
+    // get the KalSeedPtrs
+    auto handle = event.getValidHandle<KalSeedPtrCollections>(_seedTag);
+    const auto& clusters = *handle;
+    if(_debug > 0) printf("[TrackResolution::%s::%s] Collection of %zu track clusters retrieved\n", __func__, moduleDescription().moduleLabel().c_str(), clusters.size());
+
+    // Go through the tracks and check for alternate track fit hypotheses using the same helix seed
+    for (const auto& cluster : clusters) {
+      if(cluster.empty()) continue;
+      if(_debug > 1) printf("  Checking cluster of %zu tracks\n", cluster.size());
+      const KalSeedPtr* seedPtr = &(cluster.front());
+      if(_makeHists) {
+        _hists.fitcon->Fill((*seedPtr)->fitConsistency());
+      }
+
+      // Resolve the overlap for each track in the cluster
+      const size_t ntrks = cluster.size();
+      for(size_t itrk = 1; itrk < ntrks; ++itrk) {
+        const KalSeedPtr testPtr = cluster.at(itrk);
```

Comment:
Usual practice is to skip bounds checking when the indices clearly cannot be out-of-bounds (as in this case). I also don't see why you need indices at all: an expression loop should do the same thing more compactly.

Reply:
I'm only using indices in the track collection to create KalSeedPtr from the handle and the index and in the track matching loop to check tracks j > i after these updates. In the loop I can switch to an iterators and `jtrk = itrk; ++jtrk` style if this is better/cleaner. Is there a better way to create the KalSeedPtrs? I'm not very familiar with making these.

Reply:
You need indices to create Ptrs.